KAFKA-14875: Implement wakeup

philipnee opened the pull request:
Interrupt the blocking methods and throw WakeupException

philipnee committed:
wakup exception

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -546,6 +543,26 @@ public ConsumerRecords<K, V> poll(long timeout) {
+                future.cancel(true);
+                throw new WakeupException();
+            }
+            // Maybe Thread.sleep?
```
There must be a better way to do this than... sleep

Hangleton commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -546,6 +543,26 @@ public ConsumerRecords<K, V> poll(long timeout) {
+
+            if (this.shouldWakeup.get()) {
+                this.shouldWakeup.set(false);
+                future.cancel(true);
```
Should `WakeupableFuture#wakeup` be used here?

philipnee replied:
you are right!

philipnee committed:
use wakeup

philipnee committed:
doesn't need to throw wakeup exception here

philipnee committed:
clean up

philipnee committed:
clean up

wip

philipnee committed:
ensure closed

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -319,27 +330,50 @@ public Map<TopicPartition, OffsetAndMetadata> committed(final Set<TopicPartition
         }
 
         final OffsetFetchApplicationEvent event = new OffsetFetchApplicationEvent(partitions);
+        ensureNoInflightBlockingFutures(event.future());
```
it's a bit of an awkward place, but we want to make sure when we set an active future, there isn't one at the moment.

philipnee committed:
Update PrototypeAsyncConsumer.java

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -130,20 +136,20 @@ public PrototypeAsyncConsumer(final ConsumerConfig config,
-                config,
-                groupRebalanceConfig,
-                logContext,
-                subscriptions,
```
I changed my indentation to 4...

gvict90us replied:
better later then never

philipnee committed:
clean up

philipnee committed:
Add a debug log

philipnee committed:
wakeup inflight futures

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -319,26 +328,52 @@ public Map<TopicPartition, OffsetAndMetadata> committed(final Set<TopicPartition
+            throw new KafkaException(e);
         } catch (InterruptedException e) {
-            throw new InterruptException(e);
+            throw new InterruptException(e.getMessage());
```
What's up with this?

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -447,6 +482,19 @@ public void close() {
+                close(timeout, false);
+            }
+        } finally {
+            closed = true;
```
It's interesting to set this only after the call to close completes or fails.

I could see setting it only after it completes to ensure it's really closed, or setting it with a CAS before calling the inner close so that multiple callers don't all try to close at once.

I'm not sure I see the rationale of doing it this way, though. If there is a reason I'm not seeing, could you add a comment explaining it so the code will be clear?

philipnee replied:
actually that's a good point. I did this because it was the original implementation and I didn't put too much thinking into this.  For the current threading model, we definitely want to avoid multiple thread trying to close the client.  Can we synchronize on this method to prevent multithreading access?

vvcephei replied:
Yes you can! Although in general, you should not use synchronized methods, but instead use a synchronized block or a Lock or something.

gvict90us replied:
could you explain why closed is boolean but shouldWakeup is AtomicBoolean ?

vvcephei reviewed (commented):
Thanks for the PR!

I had just a couple of questions. Otherwise, it LGTM.

philipnee committed:
synchronize on close

synchronize on close.

philipnee commented:
Thanks, @vvcephei - I addressed your comments there!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/WakeupableFuture.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Can be interrupted by calling {@link #wakeup()}.
+ */
+public class WakeupableFuture<T> extends CompletableFuture<T> {
```
This class feels like overkill. The direct call to `completeExceptionally` seems clear already.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -250,6 +257,7 @@ public void commitAsync(OffsetCommitCallback callback) {
 
     @Override
     public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCommitCallback callback) {
+        maybeWakeup();
```
Why are we adding this? I don't think wakeup is expected to be thrown from async APIs?

philipnee replied:
I think you are right, it is not expected.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -479,18 +530,42 @@ public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets) {
+            throw new org.apache.kafka.common.errors.TimeoutException(e);
+        } catch (WakeupException e) {
+            this.shouldWakeup.set(false);
+            this.activeFutures.remove(commitFuture);
```
this probably should be in the finally

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -319,26 +327,52 @@ public Map<TopicPartition, OffsetAndMetadata> committed(final Set<TopicPartition
-            throw new KafkaException(e);
-        } catch (Exception e) {
+        } catch (WakeupException e) {
+            this.activeFutures.remove(event.future());
```
same, we should probably remove the future upon any failure/completion.

philipnee committed:
Addressed comments

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -546,6 +619,22 @@ public ConsumerRecords<K, V> poll(long timeout) {
+        if (this.closed)
+            throw new IllegalStateException("This consumer has already been closed.");
+
+        log.debug("Raising WakeupException in response to user wakeup");
```
Not sure this is super useful. Maybe we could add the wakeup exception so that at least we get the stack trace?

philipnee replied:
This is actually in the original wakeup call see: (I forgot why I left it out of the getAndSet block, but I'll move it into the block.
```
        if (!wakeupDisabled.get() && wakeup.get()) {
            log.debug("Raising WakeupException in response to user wakeup");
            wakeup.set(false);
            throw new WakeupException();
        }
```

hachikuji replied:
Ok. Not a big deal either way.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -546,6 +619,22 @@ public ConsumerRecords<K, V> poll(long timeout) {
 
+    private void maybeWakeup() {
+        if (this.closed)
+            throw new IllegalStateException("This consumer has already been closed.");
```
Original consumer doesn't do this. Is there a good reason to change the behavior?

philipnee replied:
there isn't and we don't need it. I'll remove this part.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -96,6 +98,11 @@
     private final Metrics metrics;
     private final long defaultApiTimeoutMs;
 
+    private final ConcurrentHashMap.KeySetView<CompletableFuture<?>, Boolean> activeFutures =
```
Is there ever a case where we have multiple futures waited on? In other words, could this be a simpler type, such as `AtomicReference`?

philipnee replied:
yeah i think so - i'll change it to atomic reference

philipnee committed:
addressed PR comments

philipnee commented:
Thanks, @hachikuji - I addressed your comment.  I took a glance at the code again, I realized there's no lock in place.  Giving it a second thought, I think we might not need it as of now. I think if two threads invoke commitSync(), one will just get blocked by the other until completion.  Is my understanding right there?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -98,8 +97,7 @@
 
-    private final ConcurrentHashMap.KeySetView<CompletableFuture<?>, Boolean> activeFutures =
-        ConcurrentHashMap.newKeySet();
+    private final AtomicReference<CompletableFuture> activeFutures = new AtomicReference<>();
```
nit: `activeFuture`?

gvict90us replied:
yes, definitely activeFuture

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -319,27 +324,53 @@ public Map<TopicPartition, OffsetAndMetadata> committed(final Set<TopicPartition
-            // Execution exception is thrown here
-            throw new KafkaException(e);
-        } catch (Exception e) {
+        } catch (WakeupException e) {
```
Hmm, wouldn't we need to unwrap this exception from `ExecutionException`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -461,6 +507,10 @@ public void close(Duration timeout) {
     public void wakeup() {
+        this.shouldWakeup.set(true);
+        if (this.activeFutures.get() != null) {
+            this.activeFutures.getAndSet(null).completeExceptionally(new WakeupException());
```
This does not seem safe. If there are concurrent calls to `getAndSet`, we might see NPE?

gvict90us commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -85,7 +87,6 @@
     private static final String CLIENT_ID_METRIC_TAG = "client-id";
     private static final String JMX_PREFIX = "kafka.consumer";
     static final long DEFAULT_CLOSE_TIMEOUT_MS = 30 * 1000;
```
well, configuration values shall be provided from outside, hard-coding is not allowed ...

gvict90us commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -96,6 +97,10 @@
+    private final AtomicReference<CompletableFuture> activeFutures = new AtomicReference<>();
+    private AtomicBoolean shouldWakeup = new AtomicBoolean(false);
+    private volatile boolean closed = false;
+
```
it looks better to have initialization in constructor but that's not a thing of first importance.

gvict90us commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -96,6 +97,10 @@
+    private AtomicBoolean shouldWakeup = new AtomicBoolean(false);
+    private volatile boolean closed = false;
+
     public PrototypeAsyncConsumer(Properties properties,
```
readability of code is below expectations ...

gvict90us commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -114,14 +119,14 @@ public PrototypeAsyncConsumer(final ConsumerConfig config,
+                ", clientId=" + clientId + ", groupId=" + groupId.orElse("null") + "] ");
         } else {
             logContext = new LogContext("[Consumer clientId=" + clientId + ", groupId=" + groupId.orElse("null") + "] ");
         }
```
a bit inconsistent: this.groupId and later groupId.orElse, when there's no need to use this then it should't be used

gvict90us commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -114,14 +119,14 @@ public PrototypeAsyncConsumer(final ConsumerConfig config,
         if (groupRebalanceConfig.groupInstanceId.isPresent()) {
             logContext = new LogContext("[Consumer instanceId=" + groupRebalanceConfig.groupInstanceId.get() +
-                    ", clientId=" + clientId + ", groupId=" + groupId.orElse("null") + "] ");
+                ", clientId=" + clientId + ", groupId=" + groupId.orElse("null") + "] ");
```
creation such of instances shall be done in separate methods

gvict90us reviewed (approved):
requested review
